File: src/AOC2022/Day2/Day2.php
<?php

namespace AOC2022\Day2;

use AOC\Day;
use AOC\Input;

enum Move {
    case ROCK;
    case PAPER;
    case SCISSOR;

    static function from(string $value): Move {
        switch ($value) {
            case 'A':
            case 'X':
                return Move::ROCK;
            case 'B':
            case 'Y':
                return Move::PAPER;
            case 'C':
            case 'Z':
                return Move::SCISSOR;
            default:
                throw new \Exception();
        }
    }

    function score(): int {
        switch ($this) {
            case Move::ROCK:
                return 1;
            case Move::PAPER:
                return 2;
            case Move::SCISSOR:
                return 3;
        }
    }

    function result(Move $opponent): Result {
        if ($opponent == $this) {
            return Result::DRAW;
        }

        switch ($opponent) {
            case Move::ROCK:
                if ($this == Move::PAPER) {
                    return Result::WIN;
                } else {
                    return RESULT::LOSS;
                }
            case Move::PAPER:
                if ($this == Move::SCISSOR) {
                    return Result::WIN;
                } else {
                    return RESULT::LOSS;
                }
            case Move::SCISSOR:
                if ($this == Move::ROCK) {
                    return Result::WIN;
                } else {
                    return Result::LOSS;
                }
        }
    }

    function roundScore(Move $opponent): int {
        return $this->score() + $this->result($opponent)->score();
    }
}

enum Result: string {
    case LOSS = 'X';
    case DRAW = 'Y';
    case WIN = 'Z';

    function score(): int {
        switch ($this) {
            case Result::LOSS:
                return 0;
            case Result::DRAW:
                return 3;
            case Result::WIN:
                return 6;
        }
    }

    function move(Move $opponent): Move {
        if ($this == Result::DRAW) {
            return $opponent;
        }

        switch ($opponent) {
            case Move::ROCK:
                if ($this == Result::LOSS) {
                    return Move::SCISSOR;
                } else {
                    return Move::PAPER;
                }
            case Move::PAPER:
                if ($this == Result::LOSS) {
                    return Move::ROCK;
                } else {
                    return Move::SCISSOR;
                }
            case Move::SCISSOR:
                if ($this == Result::LOSS) {
                    return Move::PAPER;
                } else {
                    return Move::ROCK;
                }
        }
    }

    function roundScore(Move $opponent): int {
        return $this->move($opponent)->roundScore($opponent);
    }
}

class Day2 extends Day {
    public function part1(Input $input) {
        return $input->lines->map(function ($pair) {
            [$opponent, $me] = explode(' ', $pair);

            return Move::from($me)->roundScore(Move::from($opponent));
        })
        ->sum();
    }

    public function part2(Input $input) {
        return $input->lines->map(function ($pair) {
            [$opponent, $result] = explode(' ', $pair);
            
            return Result::from($result)->roundScore(Move::from($opponent));
        })
        ->sum();
    }
}

File: swift/Sources/AOC2022/Day2.swift
import AOCKit

enum Move {
    case rock
    case paper
    case scissor
    
    init?(_ string: String) {
        switch string {
        case "A", "X": self = .rock
        case "B", "Y": self = .paper
        case "C", "Z": self = .scissor
        default: return nil
        }
    }
    
    var score: Int {
        switch self {
        case .rock: return 1
        case .paper: return 2
        case .scissor: return 3
        }
    }
    
    func result(againstOpponent opponent: Move) -> GameResult {
        switch (self, opponent) {
        case (.paper, .paper), (.rock, .rock), (.scissor, .scissor): return .draw
        case (.paper, .rock): return .win
        case (.paper, .scissor): return .loss
        case (.rock, .paper): return .loss
        case (.rock, .scissor): return .win
        case (.scissor, .paper): return .win
        case (.scissor, .rock): return .loss
        }
    }
    
    func roundScore(againstOpponent opponent: Move) -> Int {
        return score + result(againstOpponent: opponent).score
    }
}

enum GameResult {
    case loss
    case draw
    case win
    
    init?(_ string: String) {
        switch string {
        case "X": self = .loss
        case "Y": self = .draw
        case "Z": self = .win
        default: return nil
        }
    }
    
    var score: Int {
        switch self {
        case .loss: return 0
        case .draw: return 3
        case .win: return 6
        }
    }
    
    func move(againstOpponent opponent: Move) -> Move {
        switch (self, opponent) {
        case (.draw, _): return opponent
        case (.loss, .rock): return .scissor
        case (.win, .rock): return .paper
        case (.loss, .paper): return .rock
        case (.win, .paper): return .scissor
        case (.loss, .scissor): return .paper
        case (.win, .scissor): return .rock
        }
    }
    
    func roundScore(againstOpponent opponent: Move) -> Int {
        return move(againstOpponent: opponent).roundScore(againstOpponent: opponent)
    }
}

public struct Day2: Solution {
    public init() {}

    public func part1(_ input: String) -> Int {
        return input
            .lines
            .compactMap { line in
                let components = line.split(separator: " ")
                
                guard let opponent = Move(String(components[0])), let me = Move(String(components[1])) else { return nil }
                
                return me.roundScore(againstOpponent: opponent)
            }
            .sum()
    }
    
    public func part2(_ input: String) -> Int {
        return input
            .lines
            .compactMap { line in
                let components = line.split(separator: " ")
                
                guard let opponent = Move(String(components[0])), let result = GameResult(String(components[1])) else { return nil }
                
                return result.roundScore(againstOpponent: opponent)
            }
            .sum()
    }
}
